File: src/main/java/org/example/backtrack/Subsets.java
package org.example.backtrack;

import java.util.ArrayList;
import java.util.List;

public class Subsets {

    // T = O(n * 2^n)
    List<List<Integer>> res = new ArrayList<>();
    List<Integer> subset = new ArrayList<>();
    public List<List<Integer>> subsets(int[] nums) {
        dfs(0, nums);
        return res;
    }

    private void dfs(int i, int[] nums) {
        if (i >= nums.length) {
            res.add(new ArrayList<>(subset));
            return;
        }

        // Decision to add nums[i]
        subset.add(nums[i]);
        dfs(i+1, nums);

        // Decision to not add nums[i]
        subset.remove(subset.size()-1);
        dfs(i+1, nums);
    }
}

File: src/main/java/org/example/backtrack/Subsets.kt
package org.example.backtrack

fun subsets(nums: IntArray): List<List<Int>> {
    val res = mutableListOf<List<Int>>()
    val sub = mutableListOf<Int>()
    val n = nums.size

    fun backtrack(i: Int) {
        if (i == n) {
            res.add(ArrayList(sub))
            return
        }

        sub.add(nums[i])
        backtrack(i + 1)

        sub.removeAt(sub.size - 1)
        backtrack(i + 1)
    }

    return res
}

fun main() {
    
}